### DISCUSSION TITLE
[Change Request] Allow plugins to edit metadata before `on_page_markdown`

### DISCUSSION THREAD

**Discussion by McDic** [mkdocs/mkdocs#3639] [Ideas]
Hello.

I was building a plugin which tracks git history of a file **on non-metadata lines**. I wanted to track only content edit history, not metadata edit history. But I realized `self.meta` is overrided by `meta.get_date` in following lines;

https://github.com/mkdocs/mkdocs/blob/7186f4ce9c88e7c445ef90fee85c065ca297a7c8/mkdocs/structure/pages.py#L202-L215

Therefore I request `mkdocs` to change this behavior. I can upload pull requests myself if you want.

**Comment by lovelydinosaur** [comment]
> Therefore I request mkdocs to change this behavior.

It's not very clear to me exactly what you're suggesting, perhaps you could provide an example from the perspective of how the docs should look/function? (Rather than from the perspective of the code)

**Reply by McDic** [reply]
Not sure how can I describe this without perspective of the code, but ok.

Some plugins might want to change metadata of the post before `on_read_source`. mkdocs is wiping out every custom change on `self.meta` when it read the raw source of markdown files. Therefore, my plugin is adding custom metadata on `on_page_markdown`.

**Comment by pawamoy** [comment]
> Some plugins might want to change metadata of the post before on_read_source.

Why would they want that? Could you describe your use-case?

> Therefore, my plugin is adding custom metadata on on_page_markdown.

Is that not the solution?

Also, I can't seem to correlate your last comment with your initial description. IIUC, you want to track changes in the contents of Markdown documents, excluding metadata, so I'm not sure to understand why you'd want to *change* the metadata.

**Reply by McDic** [reply]
Sorry for my late reply.

My plugin is adding last edit date and created date of each document as page metadata via `git log --format "%H %ad" --date=iso-strict -L "11:20"` or something similar. Those metadatas are rendered in page later. However, the last edit date only seeks updates from non-metadata lines of the post markdown, and I don't wanted to read same page source twice.

That's why I wanted to edit custom metadata on `on_read_source`, but it's impossible, so I am doing that on `on_page_markdown`.

**Comment by kamilkrzyskow** [comment]
Git history is in the file system. It's not affected by changes in the event cycle internal to MkDocs, unless you write the changes to the file system in your plugin. 
Therefore I'm not sure if the metadata change is the issue here. 
Perhaps provide a more specific example?

**Reply by McDic** [reply]
Sorry for my late reply.

I want to scan Git history of only specific lines, which requires reading file sources. Please read [this comment](https://github.com/mkdocs/mkdocs/discussions/3639#discussioncomment-9530580) for more details.
